### ISSUE TITLE
matching: high throughput task list 

### ISSUE THREAD

**Issue by venkat1109** [uber/cadence#2098]
Cadence tasks lists are light weight task queues that are used to dispatch activity and decision task to cadence workers. Task lists are specific to a cadence domain but can be shared by multiple workflows within a domain. The throughput of a task list is directly proportional to the number of events generated per sec by the workflows that use that task list. Internally, task lists are owned and managed by cadence-matching service. Within matching service, task list ownership is determined by consistent hashing and the host responsible for a task list acquires a lease on the task list using CAS primitives at the persistence layer.  Given this architecture, it’s clear that a single task list throughput cannot scale beyond a point.

While this limitation can be documented and also easily worked around by using multiple task lists within a domain, there is no automatic way to prevent customers from accidentally relying on this limitation. The goal here is to provide a single task list offering for cadence customers that can provide high throughput / low latency without needing any code changes from customers.

**Comment by venkat1109** [comment]
Proposal
--

**High Level Overview**
On a high level, each task list will have N partitions behind the scenes where N is configurable. When history needs to add a task to a task list, a partition is picked by task_id % N at the routing / hashing layer. When workers poll for a task list, frontend will randomly pick a partition and forward the poll request to that partition. Rate limits at task list level will be split equally among all the partitions. Sync match will continue to work as is on each of the individual task list partitions. Except for the hashing / routing layer used to talk to matching service, the rest of the system will work as is unaware of the partitioning mechanism. 

**Optimization for Low Throughput Scenarios**
One problem with the approach described above is that it adds additional latency for task dispatch when throughput is low and there are multiple partitions. For example, let’s say there are two partitions on a task list; partition-0 has a backlog of 5 tasks and partition-1 has none. Now, when a poller arrives at frontend, it picks a random partition, say partition-1 and forwards the poller to the host owning that task list. Since partition-1 has no tasks, we can either block until poll timeout or immediately return and have frontend try another partition. Neither of these approaches are ideal. To solve this problem, we organize the task lists into a k-ary tree where k is configurable. Basic idea is that when addTask or pollTask arrives at a non-root partition and  the partition is idle, then forward the request to the parent with some rate limiting. Specifically, the approach requires  the following rules:

*AddTask (partition int, task Task)*
- Try local sync match first
- If local sync match fails and if partition has a parent, try doing a remote sync match for this task on parent next. Limit this type of remote sync match to at most N outstanding per task list. In addition, also rate limit at max x reqs/sec. The remote sync match returns immediately when sync match is not possible. 
- If remote sync match fails, try writing the task to the database

*Poll (partition int)*
- Try local sync match and/or drain local backlog first
- When that fails and if partition has a parent, try stealing a task from the parent next by forwarding the poll to the parent. Limit this type of remote forwarding to at most 1 outstanding per task list


*TaskReader (partition int)*
- Consume backlog from DB and try local sync match first
- If local sync match fails and if there is no outstanding remote sync match:
```
Loop:  
 Try remote_sync_match with short timeout (2s)   
 If success, we are done   
 Try local_sync_match    
 If success, we are done
```

**Failure and Other Scenarios FAQ Style**

*Q: What happens when workers are down and new tasks arrive at full speed ?*
Non-root partitions will attempt remote sync match at full speed. However, as mentioned before, remote sync match calls to parent will be throttled heavily. So, this scenario does not result in root partition getting overloaded. All of the partitions will concurrently accept new tasks at their configured rate and write them to persistence. 

*Q: What happens when new tasks arrive at full speed and everything is sync matched ?*
This is the best case where new tasks arriving in all partitions will get sync matched to their local pollers almost all of the time.

*Q: What happens when there is a backlog in all partitions and there are not enough pollers ?*
Backlog will drain slowly. Ack levels of non-leaf task list partitions might move a bit slower than ideal because ack manager may be blocked by a remote sync match attempt initiated by the task reader. However, “slow” here refers to lagging behind by a second or two in the worst case. 

*Q: What happens when new task arrives on a non-root partition and its poller is blocked on the parent ?*
For a poller to be blocked on the parent, this partition shouldn’t have had any tasks at all when that poller arrived. So, this new task that arrived will be the first task for this task list. In this case, the addTask call will attempt remote sync match to parent, which in turn will be matched to the poller which got forwarded from this partition earlier.

*Q: What happens when pollers consume tasks at roughly half the task arrival rate ?*
Backlog builds on all partitions

*Q: What happens when root partition is down ?*
If there are enough pollers overall, there will be no impact to the remaining partitions since there will be at least one poller attached to each partition. However, when there are too few pollers and the task list throughput is very low, there will be additional task dispatch latency. This is because poller might be blocked on partition X while task arrives no partition Y.
